lib/evaluate/traversal.h

```diff
@@ -19,22 +19,54 @@
 #include <type_traits>
 
 // Implements an expression traversal utility framework.
+// See fold.cc to see how this framework is used to implement detection
+// of constant expressions.
+//
+// To use, define one or more client visitation classes of the form:
+//   class MyVisitor : public virtual TraversalBase<RESULT> {
+//     explicit MyVisitor(ARGTYPE);  // single-argument constructor
+//     void Handle(const T1 &);  // callback for type T1 objects
+//     void Pre(const T2 &);  // callback before visiting T2
+//     void Post(const T2 &);  // callback after visiting T2
+//     ...
+//   };
+// RESULT should have some default-constructible type.
+// Then instantiate and construct a Traversal and its embedded MyVisitor via:
+//   Traversal<RESULT, MyVisitor, ...> t{value};  // value is ARGTYPE &&
+// and call:
+//   RESULT result{t.Traverse(topLevelExpr)};
+// Within the callback routines (Handle, Pre, Post), one may call
+//   void Stop();  // to end traversal
+//   void Return(...);  // to emplace-construct the result and end traversal
+//   RESULT &result();  // to reference the result to define or update it
+// For any given expression object type T for which a callback is defined
+// in any visitor class, the callback must be distinct from all others.
+// Further, if there is a Handle(const T &) callback, there cannot be a
+// Pre() or a Post().
+
 namespace Fortran::evaluate {
 
 template<typename RESULT> class TraversalBase {
 public:
   using Result = RESULT;
-  // Note the weird return type; it distinguishes this default Handle
-  // from any void-valued override.
+
+  Result &result() { return result_; }
+
+  // Note the odd return type; it distinguishes these default callbacks
+  // from any void-valued client callback.
   template<typename A> std::nullptr_t Handle(const A &) { return nullptr; }
-  template<typename A> void Pre(const A &) {}
-  template<typename A> void Post(const A &) {}
-  template<typename... A> void Return(A &&... x) {
-    result_.emplace(std::move(x)...);
+  template<typename A> std::nullptr_t Pre(const A &) { return nullptr; }
+  template<typename A> std::nullptr_t Post(const A &) { return nullptr; }
+
+  void Stop() { done_ = true; }
```

Comment:
I think `Return()` would be a better name for this function, to make it clear it does the same thing as the other return, except for setting the result.